### PR TITLE
T2D-236 phenotype description and type addition

### PR DESCRIPTION
**Pull request by selvaebi**

### CODE REVIEW

File: src/main/java/uk/ac/ebi/ampt2d/registry/entities/Phenotype.java [hunk 1]
```diff
@@ -47,28 +48,67 @@
...
+
+    @ApiModelProperty(position = 5, required = true)
+    @JsonProperty
+    @NotNull
+    @Column(nullable = false)
+    private String allowedValues;
+
     @CreatedDate
     @Column(updatable = false)
+    @Convert(converter = ZonedDateAttributeConverter.class)
```
Comment (Zhicheng-Liu) [inline src/main/java/uk/ac/ebi/ampt2d/registry/entities/Phenotype.java]: Have you tried `@Temporal(TIMESTAMP)` annotation? I believe that will convert between `Date` and `TIMESTAMP` in database.

File: src/main/java/uk/ac/ebi/ampt2d/registry/entities/Phenotype.java [hunk 2]
```diff
@@ -47,28 +48,67 @@
...
+    @NotNull
+    @Enumerated(EnumType.STRING)
+    @Column(nullable = false)
+    private Type type;
+
+    @ApiModelProperty(position = 5, required = true)
+    @JsonProperty
+    @NotNull
+    @Column(nullable = false)
+    private String allowedValues;
```
Comment (Zhicheng-Liu) [inline src/main/java/uk/ac/ebi/ampt2d/registry/entities/Phenotype.java:91]: What is `allowedValues` used for?

File: src/main/java/uk/ac/ebi/ampt2d/registry/entities/Property.java [hunk 3]
```diff
@@ -97,11 +98,23 @@
...
+    @Convert(converter = ZonedDateAttributeConverter.class)
     private ZonedDateTime lastModifiedDate;
 
+    public Property() {
+    }
+
+    public Property(String id, Type type, Meaning meaning, String description) {
+        this.id = id;
+        this.type = type;
+        this.meaning = meaning;
```
Comment (Zhicheng-Liu) [inline src/main/java/uk/ac/ebi/ampt2d/registry/entities/Property.java:115]: I now wonder what `meaning` is and what is it different from `description`. Are they essentially the same thing?

**Review by Zhicheng-Liu — CHANGES_REQUESTED** [review]

File: src/main/java/uk/ac/ebi/ampt2d/registry/entities/Property.java (hunk 3 above)
Comment (selvaebi) [reply src/main/java/uk/ac/ebi/ampt2d/registry/entities/Property.java:115]: I think it should be description but , meaning mostly they use it for portal side in broad.

File: src/main/java/uk/ac/ebi/ampt2d/registry/entities/Phenotype.java (hunk 2 above)
Comment (selvaebi) [reply src/main/java/uk/ac/ebi/ampt2d/registry/entities/Phenotype.java:91]: like if dichotomous 0 or 1, continuous is then positive or negative or float, multichotomous then what are the allowed values for the phenotype.

File: src/main/java/uk/ac/ebi/ampt2d/registry/entities/Property.java (hunk 3 above)
Comment (Zhicheng-Liu) [reply src/main/java/uk/ac/ebi/ampt2d/registry/entities/Property.java:115]: I think we need to clarify on this. If it is redudant we should make changes.

File: src/main/java/uk/ac/ebi/ampt2d/registry/entities/Phenotype.java (hunk 2 above)
Comment (Zhicheng-Liu) [reply src/main/java/uk/ac/ebi/ampt2d/registry/entities/Phenotype.java:91]: Is there going to be a lot of phenotypes which are multichotomous? I wonder if we can enumerate the allowed values.

Comment (selvaebi) [reply src/main/java/uk/ac/ebi/ampt2d/registry/entities/Phenotype.java:91]: I thought of making enumerated but we have continuous values which in the phenotype excel represent is n.n nn.nn kind of things so i left to be string.

File: src/main/java/uk/ac/ebi/ampt2d/registry/entities/Phenotype.java (hunk 1 above)
Comment (selvaebi) [reply src/main/java/uk/ac/ebi/ampt2d/registry/entities/Phenotype.java]: we can use it for util. Date and Calendar only i guess

Comment (Zhicheng-Liu) [reply src/main/java/uk/ac/ebi/ampt2d/registry/entities/Phenotype.java]: Or alternatively, use `java.sql.Timestamp` directly as the type for the creation date?

Comment (selvaebi) [reply src/main/java/uk/ac/ebi/ampt2d/registry/entities/Phenotype.java]: as we have the registry shared between two timezones , we used this ZonedDateTime

File: src/main/java/uk/ac/ebi/ampt2d/registry/entities/Phenotype.java (hunk 2 above)
Comment (cyenyxe) [reply src/main/java/uk/ac/ebi/ampt2d/registry/entities/Phenotype.java:91]: I would suggest renaming it as `allowedMultichtomousValues` for the sake of clarity.

File: src/main/java/uk/ac/ebi/ampt2d/registry/entities/Property.java (hunk 3 above)
Comment (cyenyxe) [reply src/main/java/uk/ac/ebi/ampt2d/registry/entities/Property.java:115]: +1 to @Zhicheng-Liu comment about clarification. One of the goals of the new implementation was to clean up the most questionable aspects of the existing API.

**Review by cyenyxe — CHANGES_REQUESTED** [review]

File: src/main/java/uk/ac/ebi/ampt2d/registry/entities/Phenotype.java (hunk 2 above)
Comment (selvaebi) [reply src/main/java/uk/ac/ebi/ampt2d/registry/entities/Phenotype.java:91]: Sorry, it is a description of allowed values, like for t2d allowedValues is 0/1, DBP its nn.nn and some multichotomous phenotype its allowedvalues etc.

File: src/main/java/uk/ac/ebi/ampt2d/registry/entities/Phenotype.java (hunk 1 above)
Comment (Zhicheng-Liu) [reply src/main/java/uk/ac/ebi/ampt2d/registry/entities/Phenotype.java]: That actually makes a case for the `Timestamp` which is independent of timezones.

Comment (selvaebi) [reply src/main/java/uk/ac/ebi/ampt2d/registry/entities/Phenotype.java]: i have used Temporal with date that solves the conversion

**Review by Zhicheng-Liu — APPROVED** [review]

File: src/test/java/uk/ac/ebi/ampt2d/registry/PropertyRegistryServiceApplicationTests.java [hunk 4]
```diff
@@ -126,7 +137,7 @@ public void shouldQueryPhenotype() throws Exception {
     public void shouldUpdatePhenotype() throws Exception {
         String location = postTestPhenotype();
 
-        mockMvc.perform(put(location).with(oAuthHelper.bearerToken("testEditor@gmail.com"))
+        mockMvc.perform(patch(location).with(oAuthHelper.bearerToken("testEditor@gmail.com"))
```
Comment (iyangar) [inline src/test/java/uk/ac/ebi/ampt2d/registry/PropertyRegistryServiceApplicationTests.java:140]: Should there be any test to update newly introduced fields

Comment (selvaebi) [reply src/test/java/uk/ac/ebi/ampt2d/registry/PropertyRegistryServiceApplicationTests.java:140]: done in 9b6b41d

**Review by iyangar — APPROVED** [review]

File: src/main/java/uk/ac/ebi/ampt2d/registry/entities/Property.java (hunk 3 above)
Comment (selvaebi) [reply src/main/java/uk/ac/ebi/ampt2d/registry/entities/Property.java:115]: Broad has asked to keep the column for now. in future they would try to remove it.

**Pull request merged**